### PR TITLE
[docs] Add more terms to the Lexicon

### PR DESCRIPTION
**Pull request by jrose-apple**
(inspired by #7987)

- canonical type
- sugared type
- generic environment
- generic signature

### CODE REVIEW

**Comment by jrose-apple** [comment]
@DougGregor, @slavapestov, feel free to improve/elaborate

@swift-ci Please smoke test and merge

File: docs/Lexicon.rst [hunk 1]
```diff
@@ -110,6 +116,20 @@ source code, tests, and commit messages. See also the `LLVM lexicon`_.
     Describes a type or function where making changes will break binary
     compatibility. See :doc:`LibraryEvolution.rst <LibraryEvolution>`.
 
+  generic environment
+    Provides context for interpreting a type that may have generic parameters
+    in it. Generic parameter types are normally just represented as "first
+    generic parameter in the outermost context" (or similar), so it's up to the
```
Comment (gottesmm) [inline docs/Lexicon.rst]: You use the phrase "outermost context" without defining it. You should be more explicit.

**Comment by gottesmm** [comment]
We should look into changing this into a MD file or something representable on github. it is unfortunate that it can not be seen via a browser on github. = /.

**Comment by jrose-apple** [comment]
Yeah. Vanilla Markdown doesn't have a "[definition list](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl#Single_term_and_description)" format, but even just using a regular list with a bit of boilerplate would at least be presentable.

**Pull request merged**